Fix parsing of sets when saveUnknown=true

dobrynin opened the pull request:
### Summary:
When `saveUnknown === true`, Dynamoose incorrectly parses sets as defined in a `Schema`. This PR resolves this issue.

### GitHub linked issue:
#516 

### Type (select 1):
- [x] Bug fix
- [ ] Feature implementation
- [ ] Documentation improvement
- [ ] Testing improvement
<!-- If you select the option below, please replace `---` below with the issue number of the GitHub issue raised, and the user who asked you to submit a broken test -->
- [ ] Test added to report bug (GitHub issue #--- @---)
- [ ] Something not listed here


### Is this a breaking change? (select 1):
- [ ] 🚨 YES 🚨
- [x] No
- [ ] I'm not sure


### Is this ready to be merged into Dynamoose? (select 1):
- [x] Yes
- [ ] No


### Are all the tests currently passing on this PR? (select 1):
- [x] Yes
- [ ] No


### Other:
- [x] I have searched through the GitHub pull requests to ensure this PR has not already been submitted
- [x] I have updated the Dynamoose documentation (if required) given the changes I made
- [x] I have added/updated the Dynamoose test cases (if required) given the changes I made
- [x] I have run `npm test` from the root of the project directory to ensure all tests continue to pass
- [x] I agree that all changes made in this pull request may be distributed and are made available in accordance with the [Dynamoose license](https://github.com/dynamoosejs/dynamoose/blob/master/LICENSE.txt)
- [x] All of my commits and commit messages are detailed, explain what changes were made, and are easy to follow and understand
- [x] I have confirmed that all my code changes are indented properly using 2 spaces
- [x] I have filled out all fields above

fishcharlie commented on lib/Attribute.js:
```diff
@@ -205,16 +205,16 @@ Attribute.prototype.setTypeFromRawValue = function(value) {
-  if (util.isArray(typeVal) || typeVal === 'list'){
-    type = 'List';
-  } else if ( (util.isArray(typeVal) && typeVal.length === 1) || typeof typeVal === 'function') {
+  if ( (util.isArray(typeVal) && typeVal.length === 1) || typeof typeVal === 'function') {
```
Will this cause merge conflicts with #521? Maybe we should ensure that we merge that PR in first, or?

dobrynin replied:
Yes it probably will cause a conflict. I think when you merge this into master you should be able to reconcile the differences?

fishcharlie commented:
@dobrynin Are we also sure this is not a breaking change? Seems very likely this could break people's code.

dobrynin commented:
@fishcharlie, I've simplified/improved the code. I think the logic I used to have could have false positives. I suppose this a breaking change because it will change the behavior of the package, although I believe it will bring it in line with what the documentation says.

I've also added tests.

fishcharlie commented:
@dobrynin 

> I suppose this a breaking change because it will change the behavior of the package, although I believe it will bring it in line with what the documentation says.

How will people need to change their code in order to fix any breaking changes this PR might cause?

fishcharlie commented:
@dobrynin Looks like we have a merge conflict

dobrynin commented:
Unfortunately I think this will be a breaking a change. Consider this example:

You have data in a DynamoDB table with a nested list field: `myList`.
You have `saveUnknown` activated.
You use this syntax for your schema:

```javascript
myMap: {
  type: 'map',
  map: {
    myList: [String]
  }
}
```

Previously this would have been handled fine. But with this change, an error will be thrown because Dynamoose will interpret `myList` as a String Set but will encounter a list when traversing the table.

fishcharlie commented:
@dobrynin Ok. Might be a while before this gets released then. Is this something you need for your own projects or something like that? I really don't want to negatively effect your software by it taking a while for this to get released.

One thing that could REALLY help get this released quicker would be writing up a detailed summary that we can include in the release notes about how people can update their code to handle this change. So like migration changes required to ensure compatibility with this change.

Also tagging @brandongoode to see if he has any input in this.

dobrynin commented:
@fishcharlie It turns out that this change actually isn't crucial for my team any longer. We've recently invested some time into updating our schemas and data (as well as Dynamoose itself) and have now been able to turn off `saveUnknown`. This PR still corrects a bug, but there is no pressure from our end to get it merged in.

As for writing up a migration document, I'd be happy to do that when I find the time.
